### PR TITLE
feat: Frequent Pattern Mining

### PR DESCRIPTION
**Pull request by tommytgraph**

### CODE REVIEW

File: algorithms/Frequent_Pattern_Mining/fpm/tg_fpm.gsql [hunk 1]
```diff
@@ -0,0 +1,147 @@
+CREATE QUERY tg_fpm(STRING unify_vertex_type, STRING item_vertex_type, STRING connect_edge, DOUBLE support, 
+                    UINT min_length, UINT lim = 10000, INT output_limit = -1, BOOL spit_to_file = FALSE,
```
Comment (victor-gsl) [inline algorithms/Frequent_Pattern_Mining/fpm/tg_fpm.gsql:2]: write_to_file is a better name than spit_to_file

File: algorithms/Frequent_Pattern_Mining/fpm/tg_fpm.gsql [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+  unify_vertex_type has a List<INT> item_list as an attribute (This can be achived by running pre_process query)
+
+  Parameters:
+  unify_vertex_type - aggregate vertex
+  item_vertex_type - item vertex
+  connect_edge - undirected edge that connect the two vertices above
+
+  support: minimum frequency to be considered
+  min_length: minimum length to be considered 
+  lim: an integer to limit the length of generated sequences.
```
Comment (victor-gsl) [inline algorithms/Frequent_Pattern_Mining/fpm/tg_fpm.gsql:19]: The names and descriptions of min_length and lim do not sufficiently describe what they are for.

File: algorithms/Frequent_Pattern_Mining/fpm/tg_fpm.gsql [hunk 3]
```diff
@@ -0,0 +1,147 @@
+CREATE QUERY tg_fpm(STRING unify_vertex_type, STRING item_vertex_type, STRING connect_edge, DOUBLE support, 
+                    UINT min_length, UINT lim = 10000, INT output_limit = -1, BOOL spit_to_file = FALSE,
+                    STRING file_path = "/home/tigergraph/tg_query_output.csv") { 
+  /*
+  Schema examples: user - (buys) - product
+                   user - (visits) - page
+                   user - (rates) - product
+
```
Comment (victor-gsl) [inline algorithms/Frequent_Pattern_Mining/fpm/tg_fpm.gsql:8]: If this algorithm follows the work of a published paper, it is good to cite the paper in the comments.

File: algorithms/Frequent_Pattern_Mining/fpm/ExprUtil.hpp [hunk 4]
```diff
@@ -0,0 +1,48 @@
+/******************************************************************************
+ * Copyright (c) 2016, TigerGraph Inc.
```
Comment (victor-gsl) [inline algorithms/Frequent_Pattern_Mining/fpm/ExprUtil.hpp:2]: Pay attention to the copyright year. You are making a change in 2021, so you should update the year.

File: algorithms/Frequent_Pattern_Mining/fpm/ExprFunctions.hpp [hunk 5]
```diff
@@ -0,0 +1,190 @@
+/*Copyright (c) 2015-2016, TigerGraph Inc.
```
Comment (victor-gsl) [inline algorithms/Frequent_Pattern_Mining/fpm/ExprFunctions.hpp:1]: Pay attention to the copyright year. You are making a change in 2021, so you should update the year.

**Review by victor-gsl — DISMISSED** [review]

File: algorithms/Frequent_Pattern_Mining/fpm/tg_fpm.gsql [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+  Parameters:
+  unify_vertex_type - aggregate vertex
+  item_vertex_type - item vertex
+  connect_edge - undirected edge that connect the two vertices above
+
+  support: minimum frequency to be considered
+  min_length: minimum length to be considered 
+  lim: an integer to limit the length of generated sequences.
+  output_limit: an integer to limit the length of the query result
```
Comment (Boyu1997) [inline algorithms/Frequent_Pattern_Mining/fpm/tg_fpm.gsql:20]: `-1` means print all, should indicate this in the comment

File: algorithms/Frequent_Pattern_Mining/fpm/tg_fpm.gsql [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+      IF (output_limit == -1 or output_limit > 0 and print_count < output_limit) THEN
+        PRINT @@frequent_pattern.top().frequency as support,
+              to_vertex_list(@@pattern_lookup.get(@@frequent_pattern.top().pattern_key)) as pattern;
+                      
+        print_count = print_count + 1;
+      END;
+
+      IF spit_to_file THEN
+        f.println(@@frequent_pattern.top().frequency, @@pattern_lookup.get(@@frequent_pattern.top().pattern_key));   
+      END;
```
Comment (Boyu1997) [inline algorithms/Frequent_Pattern_Mining/fpm/tg_fpm.gsql:142]: having `output_limit` for pint result makes sense,
not sure why a user might want to limit the output to file, maybe we should always output the entire result to file?

**Review by Boyu1997 — APPROVED** [review]
Some styling changes suggested; the overall algorithm looks good

File: algorithms/Frequent_Pattern_Mining/fpm/tg_fpm.gsql (hunk 1 above)
Comment (Boyu1997) [reply algorithms/Frequent_Pattern_Mining/fpm/tg_fpm.gsql:2]: use `file_path=""` to indicate no file output
we can remove `spit_to_file `

File: algorithms/Frequent_Pattern_Mining/fpm/ExprUtil.hpp (hunk 4 above)
Comment (Boyu1997) [reply algorithms/Frequent_Pattern_Mining/fpm/ExprUtil.hpp]: updated

File: algorithms/Frequent_Pattern_Mining/fpm/ExprFunctions.hpp (hunk 5 above)
Comment (Boyu1997) [reply algorithms/Frequent_Pattern_Mining/fpm/ExprFunctions.hpp]: updated

File: algorithms/Frequent_Pattern_Mining/fpm/tg_fpm.gsql (hunk 1 above)
Comment (Boyu1997) [reply algorithms/Frequent_Pattern_Mining/fpm/tg_fpm.gsql]: updated

File: algorithms/Frequent_Pattern_Mining/fpm/tg_fpm.gsql (hunk 2 above)
Comment (Boyu1997) [reply algorithms/Frequent_Pattern_Mining/fpm/tg_fpm.gsql]: updated

File: algorithms/Frequent_Pattern_Mining/fpm/tg_fpm.gsql [hunk 8]
```diff
@@ -0,0 +1,147 @@
+CREATE QUERY tg_fpm(STRING unify_vertex_type, STRING item_vertex_type, STRING connect_edge, DOUBLE support, 
+                    UINT min_length, UINT lim = 10000, INT output_limit = -1, BOOL spit_to_file = FALSE,
+                    STRING file_path = "/home/tigergraph/tg_query_output.csv") {
```
[the remark this answers is not in the document]
Comment (Boyu1997) [reply algorithms/Frequent_Pattern_Mining/fpm/tg_fpm.gsql]: updated

File: algorithms/Frequent_Pattern_Mining/fpm/tg_fpm.gsql (hunk 6 above)
Comment (Boyu1997) [reply algorithms/Frequent_Pattern_Mining/fpm/tg_fpm.gsql]: updated

File: algorithms/Frequent_Pattern_Mining/fpm/tg_fpm.gsql (hunk 3 above)
Comment (Boyu1997) [reply algorithms/Frequent_Pattern_Mining/fpm/tg_fpm.gsql:12]: added reference

**Pull request merged**